### PR TITLE
[MeshingApplication] Saving new mesh in a .mdpa file after the simulation run in mmg_process

### PR DESCRIPTION
**Pull request by federicoparisi**
Implementation of the remeshing after the simulation.

**Description**
The implementation regards the generation of a new mesh after the whole simulation run. Then it will save the mesh in a .mdpa file.

Please mark the PR with appropriate tags: 
- Application, finalstep

- Added final remeshing to the ExecuteFinalize in the mmg_process
- Saving in .mdpa file the mesh
- Added the dummy function of the avarage computation 

**Additional info**
The flag for activating/deactivating the option is called "remesh_post_process" and the file name is called "output_file_name". Both of them are defined in the deafult parameters in the mmg_process.py script.

### CODE REVIEW

**Review by marcnunezc — CHANGES_REQUESTED** [review]
Blocking PR, will review as soon as possible

**Review by loumalouomega — COMMENTED** [review]
Ok from my side

**Review by philbucher — CHANGES_REQUESTED** [review]
I also have a comment but no time today, will review in the next days

File: applications/MeshingApplication/python_scripts/mmg_process.py [hunk 1]
```diff
@@ -155,6 +155,8 @@ def __init__(self, Model, settings ):
             "save_external_files"              : false,
             "save_colors_files"                : false,
             "save_mdpa_file"                   : false,
+            "remesh_post_process"              : true,
```
Comment (marcnunezc) [inline applications/MeshingApplication/python_scripts/mmg_process.py]: I think we need a better name here:

```suggestion
            "remesh_at_finalize"              : true,
```

Also we lack another boolean to control the output of the mesh!

File: applications/MeshingApplication/python_scripts/mmg_process.py [hunk 2]
```diff
@@ -155,6 +155,8 @@ def __init__(self, Model, settings ):
...
+            "output_file_name"                 : "final_refined_mesh",
```
Comment (marcnunezc) [inline applications/MeshingApplication/python_scripts/mmg_process.py]:
```suggestion
            "output_mesh_file_name"                 : "final_refined_mesh",
```

File: applications/MeshingApplication/python_scripts/mmg_process.py [hunk 3]
```diff
@@ -417,12 +419,38 @@ def ExecuteInitializeSolutionStep(self):
...
                                 self.time = 0.0  # Reset
 
+    def ExecuteFinalize(self):
+        """ This method is executed in order to finalize the simulation and save the refined mesh in a new .mdpa file
+
+        Keyword arguments:
+        self -- It signifies an instance of a class.
+        """
+        self.average_remeshing = self.settings["remesh_post_process"].GetBool()
+        output_file_path = self.settings["output_file_name"].GetString()
```
Comment (marcnunezc) [inline applications/MeshingApplication/python_scripts/mmg_process.py]:
```suggestion
        output_mesh_file_name= self.settings["output_mesh_file_name"].GetString()
```

File: applications/MeshingApplication/python_scripts/mmg_process.py [hunk 4]
```diff
@@ -417,12 +419,38 @@ def ExecuteInitializeSolutionStep(self):
...
+        if self.average_remeshing:
```
Comment (marcnunezc) [inline applications/MeshingApplication/python_scripts/mmg_process.py]:
```suggestion
        self.remesh_at_finalize= self.settings["remesh_at_finalize"].GetBool()
        output_file_path = self.settings["output_file_name"].GetString()
        if self.remesh_at_finalize:
```
There is a mix of variables here. This boolean is controlling if there is a refinement at the end of the simulation (ExecuteFinalize()), so the variable `self.average_remeshing` is not accurate.

File: applications/MeshingApplication/python_scripts/mmg_process.py [hunk 5]
```diff
@@ -417,12 +419,38 @@ def ExecuteInitializeSolutionStep(self):
...
+            self._ExecuteRefinement()
+            if self.main_model_part.HasSubModelPart("fluid_computational_model_part"):
+                self.main_model_part.RemoveSubModelPart("fluid_computational_model_part")
```
Comment (marcnunezc) [inline applications/MeshingApplication/python_scripts/mmg_process.py]: This is needed to make sure that the model_part that we output is ready to be run again, and in some cases we need to remove some model parts.

However, this is too specific for the FluidDynamicsApplications so we need to write this in a more general way.

```suggestion
            for sub_model_part_name in self.sub_model_part_names_to_remove:
                if self.main_model_part.HasSubModelPart(sub_model_part_name):
                    self.main_model_part.RemoveSubModelPart(sub_model_part_name)
```

Where `self.sub_model_part_names_to_remove` is a string array written in the settings.

File: applications/MeshingApplication/python_scripts/mmg_process.py [hunk 6]
```diff
@@ -417,12 +419,38 @@ def ExecuteInitializeSolutionStep(self):
...
+            KratosMultiphysics.ModelPartIO(output_file_path, KratosMultiphysics.IO.WRITE | KratosMultiphysics.IO.MESH_ONLY).WriteModelPart(self.main_model_part)
```
Comment (marcnunezc) [inline applications/MeshingApplication/python_scripts/mmg_process.py]: Here we lack another boolean, to control if the mesh is going to be written or not. Take into account that we might want to remesh the mesh at the final step but no necessarily write it to a file.

```suggestion
            if self.output_final_mesh
                KratosMultiphysics.ModelPartIO(output_file_path, KratosMultiphysics.IO.WRITE | KratosMultiphysics.IO.MESH_ONLY).WriteModelPart(self.main_model_part)
```

Also to be read from the settings.

File: applications/MeshingApplication/python_scripts/mmg_process.py [hunk 7]
```diff
@@ -417,12 +419,38 @@ def ExecuteInitializeSolutionStep(self):
...
+        """ This method is executed in order to finalize the simulation and save the refined mesh in a new .mdpa file
+
+        Keyword arguments:
+        self -- It signifies an instance of a class.
+        """
+        self.average_remeshing = self.settings["remesh_post_process"].GetBool()
+        output_file_path = self.settings["output_file_name"].GetString()
+        if self.average_remeshing:
+            self._ExecuteRefinement()
+            if self.main_model_part.HasSubModelPart("fluid_computational_model_part"):
```
Comment (rubenzorrilla) [inline applications/MeshingApplication/python_scripts/mmg_process.py]: You can't assume the model part name is always that one.

File: applications/MeshingApplication/python_scripts/mmg_process.py [hunk 8]
```diff
@@ -417,12 +419,38 @@ def ExecuteInitializeSolutionStep(self):
                                 self.step = 0  # Reset
                                 self.time = 0.0  # Reset
 
+    def ExecuteFinalize(self):
+        """ This method is executed in order to finalize the simulation and save the refined mesh in a new .mdpa file
+
+        Keyword arguments:
+        self -- It signifies an instance of a class.
+        """
+        self.average_remeshing = self.settings["remesh_post_process"].GetBool()
```
Comment (rubenzorrilla) [inline applications/MeshingApplication/python_scripts/mmg_process.py]: Yo ease going through the code we normally keep the json name for the corresponding member vars too

File: applications/MeshingApplication/python_scripts/mmg_process.py [hunk 9]
```diff
@@ -417,12 +419,38 @@ def ExecuteInitializeSolutionStep(self):
...
+        output_file_path = self.settings["output_file_name"].GetString()
+        if self.average_remeshing:
+            self._ExecuteRefinement()
+            if self.main_model_part.HasSubModelPart("fluid_computational_model_part"):
+                self.main_model_part.RemoveSubModelPart("fluid_computational_model_part")
+            KratosMultiphysics.ModelPartIO(output_file_path, KratosMultiphysics.IO.WRITE | KratosMultiphysics.IO.MESH_ONLY).WriteModelPart(self.main_model_part)
+
+    def _compute_average_quantity(self):
```
Comment (rubenzorrilla) [inline applications/MeshingApplication/python_scripts/mmg_process.py]: Why is this required?

**Review by marcnunezc — CHANGES_REQUESTED** [review]
Hi! Find below my suggestions! I think we need to define a few more settings and write some parts more generally.

**Review by rubenzorrilla — CHANGES_REQUESTED** [review]
Don't know why my review wasn't submitted. Here my comments.

**Review by philbucher — DISMISSED** [review]
ok from my side

File: applications/MeshingApplication/python_scripts/mmg_process.py (hunk 1 above)
Comment (federicoparisi) [reply applications/MeshingApplication/python_scripts/mmg_process.py]: I changed the name, and added the boolean

File: applications/MeshingApplication/python_scripts/mmg_process.py (hunk 5 above)
Comment (federicoparisi) [reply applications/MeshingApplication/python_scripts/mmg_process.py]: This part has been modified in a more general one, adding the string in the default parameters

File: applications/MeshingApplication/python_scripts/mmg_process.py (hunk 7 above)
Comment (federicoparisi) [reply applications/MeshingApplication/python_scripts/mmg_process.py]: I modified that part, adding the string in the default parameters

File: applications/MeshingApplication/python_scripts/mmg_process.py (hunk 9 above)
Comment (federicoparisi) [reply applications/MeshingApplication/python_scripts/mmg_process.py]: This is required in order to compute the average quantity that will be used for the hessian remeshing in order to refine the mesh

Comment (philbucher) [reply applications/MeshingApplication/python_scripts/mmg_process.py]: but the function is empty ?

File: applications/MeshingApplication/python_scripts/mmg_process.py [hunk 10]
```diff
@@ -417,12 +422,42 @@ def ExecuteInitializeSolutionStep(self):
...
+        """
+        pass
+
     def ExecuteFinalizeSolutionStep(self):
         """ This method is executed in order to finalize the current step
 
         Keyword arguments:
         self -- It signifies an instance of a class.
         """
+        self.average_remeshing = self.settings["remesh_post_process"].GetBool()
```
Comment (philbucher) [inline applications/MeshingApplication/python_scripts/mmg_process.py]:
```suggestion
        self.average_remeshing = self.settings["average_remeshing"].GetBool()
```

File: applications/MeshingApplication/python_scripts/mmg_process.py [hunk 11]
```diff
@@ -417,12 +422,42 @@ def ExecuteInitializeSolutionStep(self):
...
+        self.output_final_mesh = self.settings["output_final_mesh"].GetBool()
+        output_mesh_file_name = self.settings["output_mesh_file_name"].GetString()
+        sub_model_part_names_to_remove = self.settings["sub_model_part_names_to_remove"].GetStringArray()
+        if self.remesh_at_finalize:
+            self._ExecuteRefinement()
+            for sub_model_part_name in sub_model_part_names_to_remove:
+                if self.main_model_part.HasSubModelPart(sub_model_part_name):
+                    self.main_model_part.RemoveSubModelPart(sub_model_part_name)
+            if self.output_final_mesh:
+                KratosMultiphysics.ModelPartIO(output_mesh_file_name, KratosMultiphysics.IO.WRITE | KratosMultiphysics.IO.MESH_ONLY).WriteModelPart(self.main_model_part)
```
Comment (marcnunezc) [inline applications/MeshingApplication/python_scripts/mmg_process.py]:
```suggestion
        if self.output_final_mesh:
            KratosMultiphysics.ModelPartIO(output_mesh_file_name, KratosMultiphysics.IO.WRITE | KratosMultiphysics.IO.MESH_ONLY).WriteModelPart(self.main_model_part)
```

Can you put this outside the `self.remesh_at_finalize` condition? In case someone does not remesh at `ExecuteFinalize` but still wants to output the final .mdpa.

File: applications/MeshingApplication/python_scripts/mmg_process.py (hunk 9 above)
Comment (marcnunezc) [reply applications/MeshingApplication/python_scripts/mmg_process.py]: The idea is to add the functionality in a next PR, so this are just empty functions.

@federicoparisi it is cleaner if we remove from this PR anything related to average/statistical quantities and add everything in a different PR.

File: applications/MeshingApplication/python_scripts/mmg_process.py (hunk 1 above)
Comment (marcnunezc) [reply applications/MeshingApplication/python_scripts/mmg_process.py]: I think you still need to remove the setting `"remesh_post_process"`, as it is not used in the code

File: applications/MeshingApplication/python_scripts/mmg_process.py [hunk 12]
```diff
@@ -417,6 +421,24 @@ def ExecuteInitializeSolutionStep(self):
...
+        """
+        self.remesh_at_finalize = self.settings["remesh_at_finalize"].GetBool()
+        self.output_final_mesh = self.settings["output_final_mesh"].GetBool()
+        output_mesh_file_name = self.settings["output_mesh_file_name"].GetString()
+        sub_model_part_names_to_remove = self.settings["sub_model_part_names_to_remove"].GetStringArray()
+        if self.remesh_at_finalize:
+            self._ExecuteRefinement()
+            for sub_model_part_name in sub_model_part_names_to_remove:
+                if self.main_model_part.HasSubModelPart(sub_model_part_name):
+                    self.main_model_part.RemoveSubModelPart(sub_model_part_name)
```
Comment (marcnunezc) [inline applications/MeshingApplication/python_scripts/mmg_process.py]:
```suggestion
            for sub_model_part_name in sub_model_part_names_to_remove:
                if self.main_model_part.HasSubModelPart(sub_model_part_name):
                    self.main_model_part.RemoveSubModelPart(sub_model_part_name)
            self._ExecuteRefinement()
```

Last suggestion, I think it is better to remove the submodelparts before entering the refinement.

**Review by marcnunezc — DISMISSED** [review]
Looks good to me now! Thanks for the effort!

**Comment by marcnunezc** [comment]
ping @rubenzorrilla

File: applications/MeshingApplication/python_scripts/mmg_process.py [hunk 13]
```diff
@@ -155,6 +155,10 @@ def __init__(self, Model, settings ):
             "save_external_files"              : false,
             "save_colors_files"                : false,
             "save_mdpa_file"                   : false,
+            "remesh_at_finalize"               : true,
+            "output_final_mesh"                : true,
```
Comment (rubenzorrilla) [inline applications/MeshingApplication/python_scripts/mmg_process.py]: IMO this should be set to `false` by default.

File: applications/MeshingApplication/python_scripts/mmg_process.py [hunk 14]
```diff
@@ -417,6 +421,24 @@ def ExecuteInitializeSolutionStep(self):
                                 self.step = 0  # Reset
                                 self.time = 0.0  # Reset
 
+    def ExecuteFinalize(self):
+        """ This method is executed in order to finalize the simulation and save the refined mesh in a new .mdpa file
+
+        Keyword arguments:
+        self -- It signifies an instance of a class.
+        """
+        self.remesh_at_finalize = self.settings["remesh_at_finalize"].GetBool()
```
Comment (rubenzorrilla) [inline applications/MeshingApplication/python_scripts/mmg_process.py]: Why do you need to save it as a member var if you are only using it within this function scope?

File: applications/MeshingApplication/python_scripts/mmg_process.py [hunk 15]
```diff
@@ -417,6 +421,24 @@ def ExecuteInitializeSolutionStep(self):
...
+        self.output_final_mesh = self.settings["output_final_mesh"].GetBool()
```
Comment (rubenzorrilla) [inline applications/MeshingApplication/python_scripts/mmg_process.py]: Same comment.

File: applications/MeshingApplication/python_scripts/mmg_process.py (hunk 13 above)
Comment (marcnunezc) [reply applications/MeshingApplication/python_scripts/mmg_process.py]: I agree, this is important @federicoparisi

**Review by rubenzorrilla — APPROVED** [review]
No more concerns from my side. Thanks.
